### DISCUSSION TITLE
count and detect objects

### DISCUSSION THREAD

**Discussion by Malic-92** [ultralytics/yolov5#8386] [Q&A] [answered]
How can yolov5 be tailored to give a count of total objects detected in an image/video?

**Comment by glenn-jocher** [comment, marked as answer]
@Malic-92 👋 Hello! Thanks for asking about **handling inference results**. YOLOv5 🚀 [PyTorch Hub](https://github.com/ultralytics/yolov5/issues/36) models allow for simple model loading and inference in a pure python environment without using `detect.py`. 

### Simple Inference Example

This example loads a pretrained YOLOv5s model from PyTorch Hub as `model` and passes an image for inference. `'yolov5s'` is the YOLOv5 'small' model. For details on all available models please see the [README](https://github.com/ultralytics/yolov5#pretrained-checkpoints). Custom models can also be loaded, including custom trained PyTorch models and their [exported](https://github.com/ultralytics/yolov5/issues/251) variants, i.e. ONNX, TensorRT, TensorFlow, OpenVINO YOLOv5 models.

```python
import torch

# Model
model = torch.hub.load('ultralytics/yolov5', 'yolov5s')  # yolov5n - yolov5x6 official model
#                                            'custom', 'path/to/best.pt')  # custom model

# Images
im = 'https://ultralytics.com/images/zidane.jpg'  # or file, Path, URL, PIL, OpenCV, numpy, list

# Inference
results = model(im)

# Results
results.print()  # or .show(), .save(), .crop(), .pandas(), etc.
results.xyxy[0]  # im predictions (tensor)

results.pandas().xyxy[0]  # im predictions (pandas)
#      xmin    ymin    xmax   ymax  confidence  class    name
# 0  749.50   43.50  1148.0  704.5    0.874023      0  person
# 2  114.75  195.75  1095.0  708.0    0.624512      0  person
# 3  986.00  304.00  1028.0  420.0    0.286865     27     tie

results.pandas().xyxy[0].value_counts('name')  # class counts (pandas)
# person    2
# tie       1
```
<img width="550" src="https://user-images.githubusercontent.com/26833433/149679662-b2021c78-59b2-4f52-b768-1e5161537c85.png">

See [YOLOv5 PyTorch Hub Tutorial](https://github.com/ultralytics/yolov5/issues/36) for details.

Good luck 🍀 and let us know if you have any other questions!

**Reply by omz-22** [reply]
Hi @glenn-jocher, how are you?

I'm trying to write multiple object detection results from a video to a file. But I need to write the detection results from each frame in the same line, returning:  name   confidence   class   xmin    ymin    xmax   ymax , of each detection.

I modified the yolov5 detector at the 'Print time' section as following :

        # Print time (inference-only)
        coordinates = (xyxy2xywh(torch.tensor(xyxy).view(1, 4))).view(-1).tolist()
        lines = (cls, *coordinates) 
        write = (('%g ' * len(lines)).rstrip() % lines)
        bbox = (write)
        
        save_file_results = str(Path(save_path).with_suffix('.csv'))
        save_file_logger = str(Path(save_path).with_suffix('.txt'))
                
        sys.stdout = open(save_file_results, 'a')
        print(f'{names[c]} {conf:.2f}' + ' ' + str(bbox))
        
        sys.stdout = open(save_file_logger, 'a')
        print(f'{s}Done. ({t3 - t2:.3f}s)')
        sys.stdout.close()
        
        LOGGER.info(f'{s}Done. ({t3 - t2:.3f}s)')

The results being written to the file referenced as "save_file_results" are considering only the detection with the highest confidence and the rest of the detections are not being saved.

What I'm getting from 2 objects detected in the same frame is:


![modified detector output png](https://user-images.githubusercontent.com/94117726/176673975-ffa9ab6c-6ed2-4a75-ab80-3c5ce66465af.png)


What I need  is:
	
![required output](https://user-images.githubusercontent.com/94117726/176674003-d07bfe70-26a2-440d-928f-dba0f1d34247.png)

From the logger output I can see that all the detections are being printed

![detector logger output](https://user-images.githubusercontent.com/94117726/176673889-df165873-9498-4601-94db-b318e5c71d2c.png)

Is there a way I can modify the detector to write all the detections in the way described?


Thanks,

**Reply by glenn-jocher** [reply]
@omz-22 two objects should not be detected in the same frame above NMS IoU threshold (default 0.45). If this is happening please submit a bug report to reproduce.

In terms of outputs, pandas() offers all the output formats you can want (i.e. to JSON, to CSV, to XML, etc.), so I would browse the pandas docs to see what you can do once the results are in pandas df format, i.e. df = results.pandas()

**Reply by omz-22** [reply]
Thanks for the feedback.

How can I get the the pandas() output when running inference on a video? I can only manage to do it on a single frame as referenced at the "Simple Inference Example".

**Reply by glenn-jocher** [reply]
per frame
